### DISCUSSION TITLE
how to do 2d array to pass as parameter to a c function

### DISCUSSION THREAD

**Discussion by peeble111** [exaloop/codon#372] [Q&A]
# work in cpython:
import ctypes
_dll_path = '/dvb/bridge_bin/predeal/REDEAL/libdds.so'
_dll = ctypes.CDLL(_dll_path)

class ddTableResults(ctypes.Structure):
....\_fields\_ = [("resTable", (ctypes.c_int * 4) * 5)]

class ddTableDeal(ctypes.Structure):
....\_fields_ = [("cards", (ctypes.c_uint * 4) * 4)]

tablep = ddTableResults()
tableDeal = ddTableDeal()
B='@`\x00\x00(A\x00\x00\x00$\x00\x00 \x07\x00\x00\x84\x08\x00\x00D8\x00\x00(\t\x00\x00\x00\x08\x00\x00\x00\x07\x00\x00\x90\x02\x00\x00\x10\x10\x00\x00\xc4`\x00\x008\x10\x00\x00\x00\x04\x00\x00\xc4B\x00\x00\x18\x10\x00\x00'
b = [ord(c) for c in B]

i = 0
for p in range(4):
....for s in range(4):
........tableDeal.cards[p][s] = b[i]+b[i+1]*256+b[i+2]*65536+b[i+3]*16777216
........i += 4

_dll.CalcDDtable(tableDeal, ctypes.byref(tablep))

strain = ('S', 'H', 'D', 'C', 'N')
print('   N  E  S  W')
i = 0
for T in tablep.resTable:
....print(f'{strain[i]}', end=' ')
....for t in T:
........print(f'{t:2d}', end=' ')
....i += 1
....print()

# how in codon i have to do ddTableResults and ddTableDeal ?
# thanks in advance
# Pierpaolo

**Comment by elisbyberi** [comment]
@peeble111 Does it work in CPython?

**Reply by peeble111** [reply]
**of course it work in cpython; here the output :** 

..N  E  S  W
S  6  7  6  7
H  6  6  6  6
D  5  8  5  8 
C  9  4  9  4 
N  8  5  8  5

**I am developing a project on bridge game and  the library is Bo Haglund "Double Dummy Solver".**

**Reply by elisbyberi** [reply]
You can make use of [Double Dummy Solver](https://github.com/dds-bridge/dds) from Codon directly. See https://github.com/exaloop/codon/blob/develop/docs/language/ffi.md
https://docs.exaloop.io/codon/interoperability/cpp

**Reply by peeble111** [reply]
**no example in doc about passing structure:**
struct ddTableDeal { unsigned int cards[4][4]; };
struct ddTableResults { int resTable[5][4]; };
int CalcDDtable(struct ddTableDeal tableDeal, struct ddTableResults *tablep)

**Reply by elisbyberi** [reply]
@peeble111 Take a look at: [Conversions between Codon and C/C++ types](https://docs.exaloop.io/codon/interoperability/cpp)

![image](https://user-images.githubusercontent.com/4147765/236679384-7155159b-77f9-49c4-952b-879193977920.png)

Next week, I will create a minimal reproducible example (MRE) demonstrating how to convert between Codon and C/C++ types.

**Reply by elisbyberi** [reply]
@peeble111 See: https://github.com/exaloop/codon/discussions/372#discussioncomment-5875595

**Comment by peeble111** [comment]
**I modified the script**
from C import CalcDDtable(List[List[int]], Ptr[List[List[int]]]) -> int

B='@`\x00\x00(A\x00\x00\x00$\x00\x00 \x07\x00\x00\x84\x08\x00\x00D8\x00\x00(\t\x00\x00\x00\x08\x00\x00\x00\x07\x00\x00\x90\x02\x00\x00\x10\x10\x00\x00\xc4`\x00\x008\x10\x00\x00\x00\x04\x00\x00\xc4B\x00\x00\x18\x10\x00\x00'
b = [ord(c) for c in B]

class ddTableDeal:
....cards:List[List[int]]

class ddTableResults:
....resTable:List[List[int]]    

tableDeal = ddTableDeal()
tablep = ddTableResults()

i = 0
for p in range(4):
....for s in range(4):
........tableDeal.cards[p][s] = b[i]+b[i+1]*256+b[i+2]*65536+b[i+3]*16777216
........i += 4

CalcDDtable(tableDeal, Ptr(tablep))

strain = ('S', 'H', 'D', 'C', 'N')
print('   N  E  S  W')
i = 0
for T in tablep.resTable:
....print(f'{strain[i]}', end=' ')
....for t in T:
.........print(f'{t:2d}', end=' ')
....i += 1
....print()

'''
codon run  -l /dvb/bridge_bin/predeal/REDEAL/libdds.so verysimple2.codon
'''

**but I got:**
verysimple2.codon:21:24-35: error: 'Ptr[T]' object has no method '__new__' with arguments (ddTableResults)

**Reply by elisbyberi** [reply]
@peeble111 Here is a minimal reproducible example (MRE):

`gcc -shared -o mre.so -fPIC mre.c`
``` c
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <stdbool.h>

#include <stdio.h>

int Integer(int a, int b)
{
    return a + b;
}

double Float(double a, double b)
{
    return a + b;
}

bool Boolean(bool a, bool b)
{
    if (a && b) {
        return true;
    }

    return false;
}

char Char(char a, char b)
{
    if (a > b) {
        return a;
    }

    return b;
}

char* String(int length, char* a)
{
    printf("%s\n", a);

    return a;
}

typedef struct {
    int length;
    char *chars;
} Str;

Str* Tuple(Str *s)
{
    printf("Print Python tuple in C: s.length: %d\n", s->length);
    printf("Print Python tuple in C: %s\n", s->chars);

    s->length = 13;
    strcpy(s->chars, "String from C");

    return s;
}

int main (void)
{
    Str s = {5, malloc(6)};
    strcpy(s.chars, "ABCDE");
    Tuple(&s);
    free(s.chars);
    return 0;
}
```

`codon run mre.py`
``` python
LIBRARY = "/ABSOLUTE_PATH/mre.so"
from C import LIBRARY.Integer(int, int) -> int as Integer
from C import LIBRARY.Float(float, float) -> float as Float
from C import LIBRARY.Boolean(bool, bool) -> bool as Boolean
from C import LIBRARY.Char(byte, byte) -> byte as Char
from C import LIBRARY.String(int, cobj) -> Ptr[byte] as String

i:int = Integer(2, 4)
print(i)

f:float = Float(1.5, 3.5)
print(f)

boolean:bool = Boolean(True, True)
print(boolean)

b: byte = Char(byte(65), byte(66))  # A, B
print(b)

length = 2
p: Ptr[byte] = String(length, 'AB'.c_str())
string = ''.join([str(Ptr[byte](p)[i]) for i in range(length)])
print(string)

@tuple # type Str in C
class String:
    length: int
    chars: Ptr[byte]

# import below "String" tuple declaration
from C import LIBRARY.Tuple(Ptr[String]) -> Ptr[String] as Tuple

s = String(4, "ABCD".c_str())
# print(s.length)
# print(s.chars)

p: Ptr[String] = Tuple(__ptr__(s))
t = Ptr[String](p)[0]

print("Print C struct in Python: s.length: ", t.length)
s = 'Print C struct in Python: s.chars: '
s += ''.join([str(Ptr[byte](t.chars)[i]) for i in range(t.length)])
print(s)
```
prints:
```
6
5
True
B
AB
AB
Print Python tuple in C: s.length: 4
Print Python tuple in C: ABCD
Print C struct in Python: s.length:  13
Print C struct in Python: s.chars: String from C
```

**Reply by peeble111** [reply]
but returning to my initial question, struggling for translate in codon this simple structure:
**struct ddTableDeal { unsigned int cards[4][4]; };**

**Reply by arshajii** [reply]
@peeble111 Codon doesn't have a way yet to express an array member in a struct/class, but that is just equivalent to 16 `int` fields. So you can have these fields explicitly or just use `Int[4 * 4 * 4]` which has the same size. Example:

```python
@tuple
class ddTableDeal:
    cards: Int[4 * 4 * 4]

    def get_card(self, idx: int):
        assert 0 <= idx < 16
        cards = self.cards
        p = Ptr[i32](__ptr__(cards).as_byte())
        return int(p[idx])
```

**Comment by peeble111** [comment]
codon run mre.codon
now with codon 1.19.1:
mre.codon:38 (18-35): error: no function 'Function.__new__' with arguments (Function[Tuple[Ptr[String]],Ptr[String]], Ptr[String])
